### PR TITLE
feat: Clickhouse Support

### PR DESCRIPTION
**Pull request by shadinaif**
feat: Clickhouse Support

### CODE REVIEW

File: futurex_openedx_extensions/helpers/clickhouse_operations.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+    try:
+        client = clickhouse_get_client(host='clickhouse', port=8123, username=username, password=password)
+    except Exception as exc:
+        raise ClickhouseClientConnectionError(f'Error getting Clickhouse client: {exc}') from exc
+
+    return client
+
+
+def escape_param(param_name: str, param_value: Any) -> str:
```
Comment (OmarIthawi) [inline futurex_openedx_extensions/helpers/clickhouse_operations.py:56]: Isn't that part of the client python funcitons?

Why are we re-implemting these essential functions?


```
client.query('SELECT * FROM {table:Identifier} WHERE date >= {v1:DateTime} AND string ILIKE {v2:String}', parameters=parameters)
```

https://github.com/ClickHouse/clickhouse-connect/blob/1550e8976fcad232f6d2ca873e91f86db8e3b7d5/clickhouse_connect/driver/query.py#L407-L412

Kindly be mindful of the maintainince cost of such functions.

File: futurex_openedx_extensions/helpers/clickhouse_operations.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    if isinstance(param_value, (int, float)):
+        return str(param_value)
+
+    if isinstance(param_value, (list, tuple)):
+        return '[' + ', '.join(escape_param(param_name, list_item) for list_item in param_value) + ']'
+
+    raise ClickhouseQueryParamsError(f'Unsupported parameter type: ({param_name}: {type(param_value)})')
+
+
+def replace_params(query: str, params: dict) -> str:
```
Comment (OmarIthawi) [inline futurex_openedx_extensions/helpers/clickhouse_operations.py:80]: Same concern in regards to `escape`:

Bind params seems to be doing the same process here:

https://github.com/ClickHouse/clickhouse-connect/blob/1550e8976fcad232f6d2ca873e91f86db8e3b7d5/clickhouse_connect/driver/query.py#L369C5-L404

**Review by OmarIthawi — CHANGES_REQUESTED** [review]
Thanks @shadinaif. I have some concerns over SQL injection and data saftey.

These endpoints will be soon open to Students which increases the attack vector.

Please let me know why we chose to implement escape and replace_params here instead of re-using the off-the-shelf tools.

File: futurex_openedx_extensions/dashboard/urls.py [hunk 3]
```diff
@@ -28,5 +28,10 @@
     ),
     re_path(r'^api/fx/statistics/v1/course_statuses/$', views.CourseStatusesView.as_view(), name='course-statuses'),
     re_path(r'^api/fx/statistics/v1/total_counts/$', views.TotalCountsView.as_view(), name='total-counts'),
+    re_path(
+        fr'^api/fx/query/v1/(?P<scope>course|user|tenant+)/(?P<slug>{CLICKHOUSE_QUERY_SLUG_PATTERN})/$',
```
[the remark this answers is not in the document]
Comment (OmarIthawi) [reply futurex_openedx_extensions/dashboard/urls.py:32]: @shadinaif consider re-using the constant (somehow) `{CLICKHOUSE_QUERY_CHOICES.join("|")}`

[the remark this answers is not in the document]
Comment (shadinaif) [reply futurex_openedx_extensions/dashboard/urls.py]: Done

File: futurex_openedx_extensions/helpers/models.py [hunk 4]
```diff
@@ -18,7 +32,342 @@ class Meta:
         verbose_name_plural = 'View Allowed Roles'
         unique_together = ('view_name', 'allowed_role')
 
-    def save(self, *args, **kwargs):  # type: ignore
+
+class ClickhouseQuery(models.Model):
+    """Model for storing Clickhouse queries"""
+    SCOPE_COURSE = 'course'
+    SCOPE_TENANT = 'tenant'
+    SCOPE_USER = 'user'
```
[the remark this answers is not in the document]
Comment (shadinaif) [reply futurex_openedx_extensions/helpers/models.py:39]: Done, but please consider that the scope is just a categorization thing. No permissions are applied to the scope. I suggest that we later add allowed-roles for every query. What do you think @OmarIthawi ?

File: futurex_openedx_extensions/helpers/models.py [hunk 5]
```diff
@@ -18,7 +32,342 @@ class Meta:
...
+    ALLOWED_PARAM_TYPES = [
+        PARAM_TYPE_DATE,
+        PARAM_TYPE_FLOAT,
+        PARAM_TYPE_INT,
+        PARAM_TYPE_LIST,
+        PARAM_TYPE_STR,
+    ]
+
+    DATE_METHOD_SEPARATOR = ','
+    DATE_METHODS: Dict[str, Callable[[int | None], str]] = {
```
[the remark this answers is not in the document]
Comment (shadinaif) [reply futurex_openedx_extensions/helpers/models.py]: done, converted into a class

File: futurex_openedx_extensions/helpers/models.py [hunk 6]
```diff
@@ -18,7 +32,342 @@ class Meta:
...
+    scope = models.CharField(max_length=16, choices=SCOPE_CHOICES)
+    slug = models.CharField(max_length=255)
+    version = models.CharField(max_length=4)
+    description = models.TextField(null=True, blank=True)
+    query = models.TextField()
+    params_config = models.JSONField(default=dict, blank=True)
+    paginated = models.BooleanField(default=True)
+    enabled = models.BooleanField(default=True)
+    modified_at = models.DateTimeField(auto_now=True)
+
```
[the remark this answers is not in the document]
Comment (shadinaif) [reply futurex_openedx_extensions/helpers/models.py]: done

File: futurex_openedx_extensions/helpers/clickhouse_operations.py (hunk 2 above)
Comment (shadinaif) [reply futurex_openedx_extensions/helpers/clickhouse_operations.py]: completely removed, using default client syntax now `WHERE something == %(variable1)s OR ....`

File: futurex_openedx_extensions/helpers/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (OmarIthawi) [reply futurex_openedx_extensions/helpers/models.py:39]: `allowed-roles for every query`

I suppose you're right! We might have platform-wide queries in which we need to make available to everyone on the internet.

Thanks for keeping a sharp-eye on the specs :), I'm happy that you don't listen to what I say 😅

File: futurex_openedx_extensions/helpers/models.py [hunk 7]
```diff
@@ -18,7 +32,342 @@ class Meta:
...
+        ).strftime('%Y-%m-%d'),
+    }
+
+    scope = models.CharField(max_length=16, choices=SCOPE_CHOICES)
+    slug = models.CharField(max_length=255)
+    version = models.CharField(max_length=4)
+    description = models.TextField(null=True, blank=True)
+    query = models.TextField()
+    params_config = models.JSONField(default=dict, blank=True)
+    paginated = models.BooleanField(default=True)
```
[the remark this answers is not in the document]
Comment (OmarIthawi) [reply futurex_openedx_extensions/helpers/models.py:72]: okay, I'm convinced

File: futurex_openedx_extensions/helpers/clickhouse_operations.py (hunk 1 above)
Comment (OmarIthawi) [reply futurex_openedx_extensions/helpers/clickhouse_operations.py:56]: thank you! `@thraxil` would be happy if he was working on this project :)

File: futurex_openedx_extensions/dashboard/urls.py [hunk 8]
```diff
@@ -28,5 +31,10 @@
     ),
     re_path(r'^api/fx/statistics/v1/course_statuses/$', views.CourseStatusesView.as_view(), name='course-statuses'),
     re_path(r'^api/fx/statistics/v1/total_counts/$', views.TotalCountsView.as_view(), name='total-counts'),
+    re_path(
+        fr'^api/fx/query/v1/(?P<scope>{QUERY_ALLOWED_SCOPES}+)/(?P<slug>{CLICKHOUSE_QUERY_SLUG_PATTERN})/$',
```
Comment (OmarIthawi) [inline futurex_openedx_extensions/dashboard/urls.py:35]: The plus sign here makes no sense to me.

Would you mind sharing why we'd need the scope to be `platform` or `platformmmmm`?

```suggestion
        fr'^api/fx/query/v1/(?P<scope>{QUERY_ALLOWED_SCOPES})/(?P<slug>{CLICKHOUSE_QUERY_SLUG_PATTERN})/$',
```

It only makes sense as a plain regexp e.g. `[a-z0-9]+` which would match all of `omar`, `omarrrrrrr` and `omarrrr0121212`

**Review by OmarIthawi — APPROVED** [review]
@shadinaif this is much better now!! Thank you for creating this so meticulously.

I have added few notes, but none of them is a blocker and I recommend we ship this to staging for @iamjazzar to work on while we make the final touches.

File: futurex_openedx_extensions/helpers/clickhouse_operations.py (hunk 1 above)
Comment (shadinaif) [reply futurex_openedx_extensions/helpers/clickhouse_operations.py]: stale, my bad. I've removed it. Thank you for the check

File: futurex_openedx_extensions/dashboard/urls.py (hunk 8 above)
Comment (shadinaif) [reply futurex_openedx_extensions/dashboard/urls.py]: great catch, thank you!

File: futurex_openedx_extensions/helpers/clickhouse_operations.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+        if offset >= max_count:
+            raise EmptyPage('Page does not exist!')
+        next_page = page + 1 if offset + page_size < max_count else None
+        query = f'{query} LIMIT {page_size} OFFSET {offset}'
+
+    try:
+        result = clickhouse_client.query(query, parameters=parameters)
+    except Exception as exc:
+        print('noooooooooooooooooooooooooooooooooooo')
```
[the remark this answers is not in the document]
Comment (shadinaif) [reply futurex_openedx_extensions/helpers/clickhouse_operations.py]: removed

File: futurex_openedx_extensions/helpers/clickhouse_operations.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    :type page: int
+    :param page_size: The page size.
+    :type page_size: int
+    :return: The results of the query.
+    :rtype: tuple[int | None, int | None, QueryResult | None]
+    """
+    max_count = None
+    next_page = None
+    print('--------------------------')
+    print(parameters)
```
[the remark this answers is not in the document]
Comment (shadinaif) [reply futurex_openedx_extensions/helpers/clickhouse_operations.py]: removed

**Comment by shadinaif** [comment]
All done, version bumped to v0.6.1. Merging now

**Pull request merged**